src/AspNet.Security.OAuth.SuperOffice/SuperOfficeAuthenticationHandler.cs

```diff
@@ -0,0 +1,196 @@
+            string contextId = await ProcessIdTokenAndGetContactIdentifierAsync(tokens, properties, identity);
+
+            if (string.IsNullOrEmpty(contextId))
+            {
+                throw new InvalidOperationException("An error occurred trying to obtain the context identifier from the current user's identity claims.");
+            }
+
+            // Add contextId to the Options.UserInformationEndpoint (https://sod.superoffice.com/{0}/api/v1/user/currentPrincipal).
+            string userInfoEndpoint = string.Format(CultureInfo.InvariantCulture, Options.UserInformationEndpoint, contextId);
+
+            // Get the SuperOffice user principal.
+            using var request = new HttpRequestMessage(HttpMethod.Get, userInfoEndpoint);
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);
+
+            using var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                                /* Status: */ response.StatusCode,
+                                /* Headers: */ response.Headers.ToString(),
+                                /* Body: */ await response.Content.ReadAsStringAsync());
+
+                throw new HttpRequestException($"An error occurred when retrieving SuperOffice user information ({response.StatusCode}). Please check if the authentication information is correct and the corresponding SuperOffice API is enabled.");
+            }
+
+            using var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
+            var context = new OAuthCreatingTicketContext(new ClaimsPrincipal(identity), properties, Context, Scheme, Options, Backchannel, tokens, payload.RootElement);
+            context.RunClaimActions();
+            await Events.CreatingTicket(context);
+            return new AuthenticationTicket(context.Principal, context.Properties, Scheme.Name);
+        }
+
+        private async Task<string> ProcessIdTokenAndGetContactIdentifierAsync(
+            [NotNull] OAuthTokenResponse tokens,
+            [NotNull] AuthenticationProperties properties,
+            [NotNull] ClaimsIdentity identity)
+        {
+            string contextIdentifier = string.Empty;
+
+            string idToken = tokens.Response.RootElement.GetString("id_token");
+
+            if (Options.SaveTokens)
+            {
+                // Save id_token as well.
+                SaveIdToken(properties, idToken);
+            }
+
+            IEnumerable<Claim>? claims = null;
+
+            if (Options.ValidateTokens)
```

Comment:
I guess it's taken from the Apple provider, but I'm not sure it's worth having an option to disable identity token validation (and it makes the code logic less clear).

@martincostello do you remember why we have such an option in the Apple provider?

Reply:
I do not.

Reply:
I would of included this option regardless what is in other providers. The reason it will remain is because it presents the opportunity for consumers to use it as a plain OAuth 2.0 provider; they can switch it off, while also providing by default the extra security check for those who desire it. Asking to remove this is not something I am willing to consider.

Reply:
> I would of included this option regardless what is in other providers.

Consistency is extremely important for me: the 1st goal of all the providers based on the generic OAuth 2.0 base handler - **that I've helped design** - has always been simplicity, whether they are developed by MSFT or the community. In a way, it's the opposite of the OIDC handler, which is way more customisable but much harder to get right.

We did very unusual things with the Apple provider because it was pretty much required - due to Apple not respecting the OIDC standard. Things are getting better and there are hooks or options we longer need, but we now have to live with them to avoid breaking changes.

> The reason it will remain is because it presents the opportunity for consumers to use it as a plain OAuth 2.0 provider; they can switch it off, while also providing by default the extra security check for those who desire it.

Validation can be tweaked much more extensively via `TokenValidationParameters`. I don't think we need another duplicate switch for that.

> Asking to remove this is not something I am willing to consider.

Let's make things crystal clear: there's currently no demand for SuperOffice support. As a SuperOffice employee, I get that it's important for you and we're willing to dedicate time to help you contribute to this repository - even if none of us is likely to use your product. 

But eventually, it's code **we** will have to maintain in the long term. Public APIs are super important to get right, so we must be sure we really need them, as removing them in future releases is always complicated. If you're not willing to adhere to that, then... 🤷‍♂️

Reply:
As implemented, it is simple. Much simpler than the apple provider. It couldn't be simpler.

You two have pulled my strings on this issue so much to this point, and to all of a sudden present your stance in this arrogant manner is not friendly or professional at all.

Reply:
> You two have pulled my strings on this issue so much to this point.

That's how open source contributions work, I'm afraid. Take a look at the massive PRs in the dotnet/runtime or dotnet/aspnetcore repos and you'll see that it's not specific to this repository: every time there's a design issue, it takes some time to come to an agreement.

> and to all of a sudden present your stance in this arrogant manner is not friendly or professional at all.

If there's anyone arrogant here, it's you: "asking to remove this is not something I am willing to consider".

If you're looking for a more professional attitude, we can definitely make you pay for the time we dedicated to this PR. You realize we do that for free, right?

Reply:
@kevinchalet Let's say I take validation out... What's further changes are required? 
Everything else is that is in ProcessIdTokenAndGetContactIdentifierAsync function must remain to obtain the contextIdentifier - which is required to construct the UserInfoEndpoint. Do you recall commenting on how I pulled the [contextId from the access_token](https://github.com/AnthonyYates/AspNet.Security.OAuth.Providers/blob/a54830b7547c660060258c79ad242cc39c65fc91/src/AspNet.Security.OAuth.SuperOffice/SuperOfficeAuthenticationHandler.cs#L47)? You had issues with that, and so I changed the code to validate the token in order to pull out the contextIdentifer claim (used to construct the UserInfoEndpoint). Now I recall I don't need validation to do that, and therefore am willing to remove it to satisfy your desire for simplicity and eliminate the ValidateAsync method.

Question becomes, will that be enough to satisfy you?

Reply:
Re-read my comment: I never suggested removing validation, I suggested removing the `ValidateTokens` option that allows disabling the validation logic... i.e the `else` block (where you parse the id_token but don't validate it).

Reply:
> I'm not sure it's worth having an option to disable identity token validation 
I completely misunderstood this comment. The context could have been made much clearer by stating the intent as you last comment has included.

I am sorry for the misunderstanding.

A fix has been issued, and tests updated. Anything else?

Reply:
> A fix has been issued, and tests updated.

I can't see them. Did you forget to `git push`?

> Anything else?

Nope. When you're done with this change, please fix the conflicts so we can merge your PR.

![image](https://user-images.githubusercontent.com/6998306/82464122-fcdc1680-9abd-11ea-902c-25a582701686.png)

Reply:
@kevinchalet Are you still observing conflicts on your side? I see this from my side. 
![image](https://user-images.githubusercontent.com/1012665/82482615-61f03600-9ad7-11ea-9881-3c91695f1646.png)

Are you able to provide and conflict details?

Reply:
Conflicts are resolved now:

![image](https://user-images.githubusercontent.com/1439341/82643809-18ebcf00-9c08-11ea-8634-dccd81aaab67.png)

Reply:
Hi @martincostello and @kevinchalet . Yes, took me a bit to realize I needed to fetch upstream to see the referenced conflicts. But now that the conflict has been done, and rebasing dev, I'm a bit nervous because I see the latest Discord changes also in my PR... I expected the Apple changes (as requested by Martin to keep it), but not the Discord changes.
Obviously still adjusting to git (from TFS), so don't want to break anything. 

![image](https://user-images.githubusercontent.com/1012665/82645491-3bcfb080-9c13-11ea-97aa-7bcfc5ef624c.png)

Best regards.

Reply:
They're the same changes as from #427 so they should be fine.

Once Kévin gives the nod, I'll squash and merge this PR and it'll go into the dev branch. You'll be able to pull a package from our MyGet feed then until we release 3.1.2, which I'm hoping to do once we've sorted out updating the CI and publish pipeline in #423.